File: swi_prolog/langford.pl
/*

  Langford's number problem L(2,N) in SWI Prolog

  Langford's number problem (CSP lib problem 24)
  http://www.csplib.org/prob/prob024/
  """
  Arrange 2 sets of positive integers 1..k to a sequence,
  such that, following the first occurence of an integer i, 
  each subsequent occurrence of i, appears i+1 indices later
  than the last. 
  For example, for k=4, a solution would be 41312432
  """
  
  * John E. Miller: Langford's Problem
    http://www.lclark.edu/~miller/langford.html
  
  * Encyclopedia of Integer Sequences for the number of solutions for each k
    http://www.research.att.com/cgi-bin/access.cgi/as/njas/sequences/eisA.cgi?Anum=014552


  Note: For k=4 there are two different solutions:
     solution:[4,1,3,1,2,4,3,2]
     position:[2,5,3,1,4,8,7,6]
  and
     solution:[2,3,4,2,1,3,1,4]
     position:[5,1,2,3,7,4,6,8]

  With this symmetry breaking

     Solution[1] < Solution[K2],

  then just the second solution is shown.

  Note: There are only solutions when K mod 4 == 0 or K mod 4 == 3.
  
  Model created by Hakan Kjellerstrand, hakank@gmail.com
  See also my SWI Prolog page: http://www.hakank.org/swi_prolog/

*/

:- use_module(library(clpfd)).
:- use_module(hakank_utils).

%%
%% Solution of K=4 (see above).
%%
go :-
    K = 4,
    format("K: ~d~n", K),
    time(langford(K, Solution, Position)),
    format("Solution: ~w~n", [Solution]),
    format("Position: ~w~n", [Position]),
    nl.

%%
%% Get the first (if any) solutions for K in 2..40
%%
go2 :- 
        between(2,60,K),
        format("K: ~d\n", [K]),
        ( time(langford(K, Solution, _Position))
        -> 
          (
           nonvar(Solution)
          -> 
           format("Solution: ~w ~n", [Solution])
          ;
            writeln("No solution")
         )
        ;
          writeln("")
        ),
        nl,
        fail,
        nl.

go2.


%%
%% Count the number of solutions
%%
go3 :- 
        between(2,11,K),
        format("k: ~d~n",[K]),        
        time(findall(1, langford(K, _Solution, _Position), L)),
        length(L,Len),
        format("K: ~d = ~d solutions~n~n", [K, Len]),
        fail,
        nl.

go3.

 
langford(K, Solution, Position) :-

        ( \+ ((K mod 4 #= 0; K mod 4 #= 3))
        -> 
          writeln("There is no solution for K unless K mod 4 == 0 or K mod 4 == 3"),
          fail
        ;
          true
        ),
        
        K2 #= 2*K,
        length(Position,K2),
        Position ins 1..K2,
        
        length(Solution,K2),
        Solution ins 1..K,
        
        all_distinct(Position),

        % symmetry breaking:
        element(1,Solution,Solution1),
        element(K2,Solution,SolutionK2),
        Solution1 #< SolutionK2,

        % main constraints
        numlist(1,K,Is),
        maplist(langford_constraints2(K,Position,Solution),Is),
        
        append(Solution,Position,Vars),
        labeling([ff,enum], Vars).

%% for maplist
langford_constraints2(K,Position,Solution,I) :-
        KI #= K+I,
        element(KI, Position,PositionKI),
        element(I, Position,PositionI),
        I1 #= I+1,
        PositionKI #= PositionI + I1,
        element(I,Position,PositionI),
        element(PositionI,Solution,I), 
        element(PositionKI,Solution,I).

File: julia/constraints/langford.jl
#=#
  Langford's number problem L(2,N) in Julia ConstraintSolver.

  Langford's number problem (CSP lib problem 24)
  http://www.csplib.org/prob/prob024/
  """
  Arrange 2 sets of positive integers 1..k to a sequence,
  such that, following the first occurence of an integer i,
  each subsequent occurrence of i, appears i+1 indices later
  than the last.
  For example, for k=4, a solution would be 41312432
  """

  * John E. Miller: Langford's Problem
    http://www.lclark.edu/~miller/langford.html

  * Encyclopedia of Integer Sequences for the number of solutions for each k
    http://www.research.att.com/cgi-bin/access.cgi/as/njas/sequences/eisA.cgi?Anum=014552


  Note: For k=4 there are two different solutions:
     solution:[4,1,3,1,2,4,3,2]
     position:[2,5,3,1,4,8,7,6]
  and
     solution:[2,3,4,2,1,3,1,4]
     position:[5,1,2,3,7,4,6,8]

  Which this symmetry breaking

     Solution[1] #< Solution[K2],

  then just the second solution is shown.

  Note: There are only solutions when K mod 4 == 0 or K mod 4 == 3.

  - For a generalized version L(k,n) where k >= 2, see http://hakank.org/langford_generalized.pi
  - For Nickerson's variant, see http://hakank.org/langford_nickerson.pi

  Model created by Hakan Kjellerstrand, hakank@gmail.com
  See also my Julia page: http://www.hakank.org/julia/


=#
using ConstraintSolver, JuMP
using Cbc, GLPK
const CS = ConstraintSolver
include("constraints_utils.jl")


#
# Note: my_element is defined in constraints_utils.jl
#
function langford(k=5,symmetry_breaking=true,all_solutions=true,print_solutions=true)

    if !((k % 4 == 0) || (k % 4 == 3))
        println("There is no solutions for k ($k) unless K mod 4 == 0 or K mod 4 == 3")
        return
    end

    cbc_optimizer = optimizer_with_attributes(Cbc.Optimizer, "logLevel" => 0)
    glpk_optimizer = optimizer_with_attributes(GLPK.Optimizer)

    model = Model(optimizer_with_attributes(CS.Optimizer,   "all_solutions"=> all_solutions,
                                                            # "all_optimal_solutions"=>true,
                                                            "logging"=>[],

                                                            "traverse_strategy"=>:BFS,
                                                            # "traverse_strategy"=>:DFS, # <-
                                                            # "traverse_strategy"=>:DBFS,

                                                            "branch_split"=>:Smallest,
                                                            # "branch_split"=>:Biggest,
                                                            # "branch_split"=>:InHalf, # <-

                                                            # "simplify"=>false,
                                                            # "simplify"=>true, # default

                                                            "time_limit"=>16,

                                                            # "lp_optimizer" => cbc_optimizer,
                                                            # "lp_optimizer" => glpk_optimizer,
                                        ))
    k2 = 2*k

    @variable(model, 1 <= position[1:k2] <= k2, Int)
    @variable(model, 1 <= solution[1:k2] <= k, Int)

    @constraint(model, position in CS.AllDifferentSet())

    # symmetry breaking
    if symmetry_breaking
        @constraint(model, solution[1] <= solution[k2])
    end

    for i in 1:k
        @constraint(model, position[k+i] == position[i] + i+1)
        # @constraint(model, solution[position[i]] == i )
        my_element(model, position[i], solution, i)
        my_element(model, position[k+i], solution, i)
    end


    # Solve the problem
    optimize!(model)

    status = JuMP.termination_status(model)
    # println("status:$status")
    if status == MOI.OPTIMAL
        num_sols = MOI.get(model, MOI.ResultCount())
        println("num_sols:$num_sols\n")
        if print_solutions
            for sol in 1:num_sols
                println("solution #$sol")
                positionx = convert.(Integer,JuMP.value.(position; result=sol))
                solutionx = convert.(Integer,JuMP.value.(solution; result=sol))
                # println("position:$positionx\n")
                println("solution:$solutionx position:$positionx\n")
            end
        end
    else
        println("status:$status")
    end

    return status
end

all_solutions = true
print_solutions = true
symmetry_breaking = true
@time langford(7,symmetry_breaking,all_solutions,print_solutions)

for n in 1:20
    if n % 4 == 3  || n % 4 == 0
        println("n:$n")
        @time status = langford(n,true,all_solutions,false)
        if status != MOI.OPTIMAL
            break
        end
    end
end

println("\nTime to first solution:")
all_solutions = false
print_solutions = true
symmetry_breaking = true
for n in 1:20
    if n % 4 == 3  || n % 4 == 0
        println("n:$n")
        @time status = langford(n,symmetry_breaking,all_solutions,print_solutions)
        if status != MOI.OPTIMAL
            break
        end
    end
end
